react-native-hub-app/src/checklist.tsx

```diff
@@ -0,0 +1,482 @@
+ */
+
+import React from 'react';
+import {
+  FlatList,
+  StyleSheet,
+  View,
+  Text,
+  TouchableOpacity,
+  AsyncStorage,
+  Linking,
+} from 'react-native';
+import {Client, Where} from '@textile/threads-client';
+import {ThreadID} from '@textile/threads-id';
+import {Buckets, Context, createAPISig} from '@textile/textile';
+import {Libp2pCryptoIdentity} from '@textile/threads-core';
+// @ts-ignore
+import {USER_API_SECRET, USER_API_KEY, API_URL} from 'react-native-dotenv';
+import styles from './styles'
+import { astronautSchema, createAstronaut } from './astronauts'
+
+const MAX_STEPS = 2;
+const version = 4
+const IDENTITY_KEY = 'identity-' + version;
+const CONTEXT_KEY = 'context-' + version;
+const TOKEN_KEY = 'token-' + version;
+const USER_THREAD_ID = 'user_thread-' + version;
+const sleep = (m) => new Promise((r) => setTimeout(r, m));
+
+interface StateProps {
+  steps: any,
+  step: number,
+  errorMessage: string,
+  threadId?: ThreadID,
+  ctx?: Context,
+  db?: Client,
+  entityId?: string,
+  bucketUrl?: string
+}
+class CheckList extends React.Component<StateProps> {
+  constructor(props) {
+    super(props);
+  }
+
+  // you could also do this, so no constructor needed
+  state: StateProps = {
+    steps: [
+      {key: 'Step 0', name: 'Prepare Identity & API Token', status: 0},
+      {key: 'Step 1', name: 'Setup ThreadDB', status: 0},
+      {key: 'Step 2', name: 'Add Instance to Collection', status: 0},
+      {key: 'Step 3', name: 'Query from our Collection', status: 0},
+      {key: 'Step 4', name: 'Push webpage to User Bucket', status: 0},
+    ],
+    step: 0,
+    errorMessage: '',
+  };
+
+  incrementStatus(index: number) {
+    const steps = this.state.steps;
+    const data = steps[index];
+    if (data.status >= MAX_STEPS) {
+      return;
+    }
+    data.status = data.status += 1;
+    steps[index] = data;
+    this.setState({
+      steps: steps,
+    });
+  }
+
+  async generateIdentity(): Promise<Libp2pCryptoIdentity> {
+    let idStr = await AsyncStorage.getItem(IDENTITY_KEY);
+    if (idStr) {
+      return await Libp2pCryptoIdentity.fromString(idStr);
+    } else {
+      const id = await Libp2pCryptoIdentity.fromRandom();
+      idStr = id.toString();
+      await AsyncStorage.setItem(IDENTITY_KEY, idStr);
+      return id;
+    }
+  }
+
+  async getUserToken(database, id): Promise<string> {
+    let token;// = await AsyncStorage.getItem(TOKEN_KEY);
+    if (token) {
+      return token;
+    }
+    token = await database.getToken(id);
+    await AsyncStorage.setItem(TOKEN_KEY, token);
+    return token;
+  }
+
+  async getContext(): Promise<Context> {
+    // Pull the stored context to reuse if available && valid date
+    let contextStr = await AsyncStorage.getItem(CONTEXT_KEY);
+    if (contextStr) {
+      const ctxJson = JSON.parse(contextStr);
+      if (
+        ctxJson['x-textile-api-sig-msg'] && (Date.parse(ctxJson['x-textile-api-sig-msg'])) > (new Date()).getTime()) {
+        // Not expired
```

Comment:
ah right... here's another one. an expired token right now can only be detected if i parse grpc error messages and _guess_ that the unauthorized means my token needs a refresh. is there a way context could help with this? i was thinking, `ctx.expired` should be boolean. perhaps it should error any api call before it even tries to make it with a specific expired error, so then a developer will learn more easily what to do.

Reply:
we could check `x-textile-api-sig-msg` each time? I _think_ right now that (i.e., 'x-textile-api-sig') is the only thing that expires, right?

Reply:
added https://github.com/textileio/js-threads/issues/186